### PR TITLE
Introduce a debug package

### PR DESCRIPTION
**Pull request by porridge**
One new exported function: `CollectArtifacts` which gathers artifacts from a k8s cluster namespace which can help debugging failures. For now, these are YAML-format dumps of all resources in that namespace.

Invokes the `kubectl` binary where this is already implemented.

Eventually this implementation will probably be replaced with the diagnostics bundle ([KEP-22](https://github.com/kudobuilder/kudo/pulls?q=is%3Apr+is%3Aopen+label%3Akep%2F22)).

### CODE REVIEW

**Comment by porridge** [comment]
Demo: [log](https://teamcity.mesosphere.io/viewLog.html?tab=buildLog&logTab=tree&filter=debug&expand=all&buildId=2676986&_focus=7308), [artifacts](https://teamcity.mesosphere.io/viewLog.html?buildId=2676986&buildTypeId=Frameworks_DataServices_Kudo_Cassandra_Pr_CassandraPrKonvoyKudo_2&tab=artifacts#!anpigxo1qish).

File: pkg/debug/debug.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+	"sync"
+
+	"github.com/spf13/afero"
+)
+
+const (
+	testArtifactsDirectoryVarName = "TEST_ARTIFACTS_DIRECTORY"
+)
+
+var (
```
Comment (nfnt) [inline pkg/debug/debug.go]: Let's not use globals. Instead, to improve the testability, we should have a struct that has these values and methods to collect the artifacts. The `CollectArtifacts` helper would then instantiate this class, call its method and return the output. The tests would then test the method.

File: pkg/debug/debug.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+	// globals for overriding in unit tests
+	artifactsDirectoryBase = os.Getenv(testArtifactsDirectoryVarName) // nolint:gochecknoglobals
+	execCommand            = exec.Command                             // nolint:gochecknoglobals
+)
+
+// CollectArtifacts collects useful debugging artifacts from a given namespace.
+// Should typically be called if CurrentGinkgoTestDescription().Failed, like this:
+//   debug.CollectArtifacts(afero.NewOsFs(), GinkgoWriter, TestNamespace, KubeConfigPath, KubectlPath)
+func CollectArtifacts(fs afero.Fs, writer io.Writer, namespace, kubeConfigPath, kubectlPath string) {
```
Comment (nfnt) [inline pkg/debug/debug.go]: We should assume that a user of `test-tools` has a `client.Client` instance and make use of that here. This would become
```suggestion
func CollectArtifacts(client client.Client, fs afero.Fs, writer io.Writer, namespace string)
```

File: pkg/debug/debug.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+	err := collectNamespacedResources(fs, writer, namespace, kubeConfigPath, kubectlPath)
+	if err != nil {
+		_, _ = fmt.Fprintf(writer, "collection of resources for debugging failed: %v\n", err)
```
Comment (nfnt) [inline pkg/debug/debug.go:49]: Please add error handling. Here and for the other cases of `fmt.Fprintf` below.

File: pkg/debug/debug.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+func collectNamespacedResources(fs afero.Fs, writer io.Writer, namespace, kubeConfigPath, kubectlPath string) error {
+	if artifactsDirectoryBase == "" {
+		return fmt.Errorf("$%s not set", testArtifactsDirectoryVarName)
+	}
+
+	_, _ = fmt.Fprintf(writer, "collecting namespaced resources for debugging...\n")
+
+	cmd := execCommand(
```
Comment (nfnt) [inline pkg/debug/debug.go]: You should use the `cmd` package of `test-tools` for that. Is there a reason why this needs to call `kubectl` or can this be done using the Kubernetes API instead? I see this as a good opportunity to add the needed types for this to our own `kubernetes` package and dogfood this here.

File: pkg/debug/debug.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+		if group == "" {
+			fileName = "resources.yaml"
+		} else {
+			fileName = fmt.Sprintf("resources-%s.yaml", group)
+		}
+
+		wg.Add(1)
+
+		go collectResources(fs, writer, &wg, namespace, fileName, resources,
```
Comment (nfnt) [inline pkg/debug/debug.go]: Is there a reason other than possible performance gains to do this in parallel? Are there performance issues when running this in serial? This looks like premature optimization and there could be issues using `afero.Fs` concurrently... are `fs` and `writer` thread-safe? Let's not use go-routines here unless there's a good reason to do so. And if there is a good reason to do so, use channels to synchronize writing to `fs` and `writer`.

File: pkg/debug/debug.go [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+func collectResources(fs afero.Fs, writer io.Writer, wg *sync.WaitGroup, namespace, fileName, resourcesNames,
+	directoryName, kubectlPath, kubeConfigPath string) {
+	defer wg.Done()
+
+	outPath := path.Join(directoryName, fileName)
+
+	output, err := fs.Create(outPath)
+	if err != nil {
+		_, _ = fmt.Fprintf(writer, "creating %q failed: %v\n", outPath, err)
```
Comment (nfnt) [inline pkg/debug/debug.go:146]: Why not return an error here? Same for the `fmt.Fprintf` below.

File: pkg/debug/debug.go [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+	return nil
+}
+
+func collectResources(fs afero.Fs, writer io.Writer, wg *sync.WaitGroup, namespace, fileName, resourcesNames,
+	directoryName, kubectlPath, kubeConfigPath string) {
+	defer wg.Done()
+
+	outPath := path.Join(directoryName, fileName)
+
+	output, err := fs.Create(outPath)
```
Comment (nfnt) [inline pkg/debug/debug.go:144]: Let's use a temporary output and only create and write the file once it has been verified that the command succeeded and the output isn't empty. This way we have a clear path that makes it easier for this function to return an error instead of writing error conditions to `io.Writer` and we don't have to remove the file if it's empty which simplifies the code.

File: pkg/debug/debug.go [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+		go collectResources(fs, writer, &wg, namespace, fileName, resources,
+			artifactsDirectory, kubectlPath, kubeConfigPath)
+	}
+
+	wg.Wait()
+
+	return nil
+}
+
+func collectResources(fs afero.Fs, writer io.Writer, wg *sync.WaitGroup, namespace, fileName, resourcesNames,
```
Comment (nfnt) [inline pkg/debug/debug.go]: I dislike the idea of using `io.Writer` to propagate error conditions. Ideally, functions with error conditions should return an error instead. Of course, there are sometimes situations where this can't be the case but I think that this function could be refactored to return an `error`.

**Review by nfnt — CHANGES_REQUESTED** [review]
This will be a great addition to aid in debugging, thanks for this! Though, I have quite a few comments, as I'd like to dogfood other `test-tools` packages as much as possible.

**Comment by porridge** [comment]
Thanks for the review @nfnt , here are a few high-level responses first:
1. the `CollectArtifacts` helper not returning an error (and therefore ignores errors explicitly) is a conscious design decision. This is because the caller (a test suite) would have nothing to do with that error anyway except display it, because when `CollectArtifacts` is called, it means the test has already failed.
2. Closely related to the above: aggregating (possibly multiple) errors into a collection and then forcing the caller to somehow iterate or otherwise display all of them would be bad UX - we want individual suites to only contain a single function call, not duplicate error handling logic. This is why we take a `Writer` and report the errors ourselves.
3. The reason for using `kubectl` rather than the API directly is mentioned in the PR description - this is really meant to be a stop-gap solution until we can use the outcome of KEP-22, and I don't want to re-implement the same logic as `kubectl` already does.
4. As for running multiple commands in parallel - I feel this optimization is very much needed, I added it as one of the last things as a result of using this package - the serial experience is pretty horrible when used interactively - one just sits there waiting for all those calls to finish.

File: pkg/debug/debug.go (hunk 1 above)
Comment (porridge) [reply pkg/debug/debug.go]: I did this, although I'm not very fond of the result: I didn't find a good name for the struct, and this introduces a method that is not covered with tests by design :-/

File: pkg/debug/debug.go (hunk 2 above)
Comment (porridge) [reply pkg/debug/debug.go]: See point 3 in https://github.com/kudobuilder/test-tools/pull/24#issuecomment-607104384 for reason why I'm using `kubectl`.

File: pkg/debug/debug.go (hunk 3 above)
Comment (porridge) [reply pkg/debug/debug.go:49]: See point 1 in https://github.com/kudobuilder/test-tools/pull/24#issuecomment-607104384 for the reason I'm swallowing the error here and elsewhere.

File: pkg/debug/debug.go (hunk 4 above)
Comment (porridge) [reply pkg/debug/debug.go]: I looked at the `cmd` package and I'm not very fond of it - it does not seem to add any value, makes some `os/exec` features unavailable, introduces its own specific syntax. I'd rather use the standard package.

See point 3 in https://github.com/kudobuilder/test-tools/pull/24#issuecomment-607104384 for why I'm using `kubectl`.

File: pkg/debug/debug.go (hunk 5 above)
Comment (porridge) [reply pkg/debug/debug.go]: See point 4 in https://github.com/kudobuilder/test-tools/pull/24#issuecomment-607104384

File: pkg/debug/debug.go (hunk 6 above)
Comment (porridge) [reply pkg/debug/debug.go:146]: See point 1 in https://github.com/kudobuilder/test-tools/pull/24#issuecomment-607104384

File: pkg/debug/debug.go (hunk 7 above)
Comment (porridge) [reply pkg/debug/debug.go:144]: By "temporary output" you mean collecting it in memory, or a temporary file elsewhere that would then be copied here?

File: pkg/debug/debug.go (hunk 8 above)
Comment (porridge) [reply pkg/debug/debug.go]: See point 2 in https://github.com/kudobuilder/test-tools/pull/24#issuecomment-607104384

**Comment by porridge** [comment]
@nfnt I think I responded to all your points, PTAL

**Comment by porridge** [comment]
This is now being tested in https://github.com/mesosphere/kudo-cassandra-operator/pull/86
[Artifacts](https://teamcity.mesosphere.io/viewLog.html?buildId=2701522&buildTypeId=Frameworks_DataServices_Kudo_Cassandra_Pr_CassandraPrKonvoyKudo_2&tab=artifacts#!ogvbc17w,-o4b5n5m2e8mv,anpigxo1qish,h8fkownehjol), [log](https://teamcity.mesosphere.io/viewLog.html?tab=buildLog&logTab=tree&filter=debug&expand=all&buildId=2701522&_focus=7627)

File: pkg/debug/debug.go (hunk 2 above)
Comment (nfnt) [reply pkg/debug/debug.go]: This isn't about `kubectl`. `client.Client` is supposed to wrap everything necessary to use the various functions provided by this library. Collecting cluster artifacts should be one of those functions. Let's add `kubeconfigPath` as an optional value to the `client.Client` struct. This function will then use this value if it is set or return an error otherwise.

File: pkg/debug/debug.go (hunk 4 above)
Comment (nfnt) [reply pkg/debug/debug.go]: What's missing in `cmd` that you need here? Would be great to add it to `cmd` then.

File: pkg/debug/debug.go (hunk 7 above)
Comment (nfnt) [reply pkg/debug/debug.go:144]: Collect it in memory.

**Comment by nfnt** [comment]
Great comments @porridge, let me address some of them. Overall, I'd love to have this functionality in `test-tools`, but we should find the right balance between following the interfaces of this library while at the same time enabling this function. From a library PoV, the function signature is my biggest concern, because we should avoid changing them later to provide backwards-compatibility. And for that we should use `client.Client` as the object that will provide the `kubectlConfigPath`. With that we keep consistency with the other packages in this library. E.g.
```go
func CollectArtifacts(client client.Client, fs afero.Fs, writer io.Writer, namespace string, kubectlPath string)
```
I'm fine with calling `kubectl` as part of this function, but please document this dependency, as the rest of the library doesn't have this dependency. I'm also fine with the actual implementation details, as possible issues (e.g. with running things in parallel) can be fixed later without changing the function signature.

tl;dr: Let's have a function signature that doesn't feel foreign to the rest of the library, the actual implementation is fine.

Some comments to your points though:
1) I disagree, let's not ignore errors. We can't assume how `CollectArtifacts` is called. It should return errors for conditions that prevent this function to work as expected. E.g. if `kubectl` command fails because `kubectlPath` was wrong. That is of interest to the caller.
2) I agree that aggregating errors here isn't good. Let's keep the `Writer` approach
3) 👍, let's use `kubectl`. I briefly looked into how this could be done using the API and agree that `kubectl` is the easiest way. Let's document this dependency though, as the rest of the library only uses `client-go` and could be used in environments without `kubectl`.
4) I still feel that this will cause trouble in the future because this isn't thread-safe usage of `fs` and `writer` but am happy to keep this parallel. We can always fix issues here in the future without changing the function signature.

**Comment by porridge** [comment]
This is now being tested in https://github.com/mesosphere/kudo-cassandra-operator/pull/86
[Artifacts](https://teamcity.mesosphere.io/viewLog.html?buildId=2704833&buildTypeId=Frameworks_DataServices_Kudo_Cassandra_Pr_CassandraPrKonvoyKudo_2&tab=artifacts#!anpigxo1qish,h8nut2u5sdyb), [log](https://teamcity.mesosphere.io/viewLog.html?tab=buildLog&logTab=tree&filter=debug&expand=all&buildId=2704833&_focus=7321)

**Comment by porridge** [comment]
Feedback addressed, PTAL @nfnt

**Review by nfnt — APPROVED** [review]
This is looking great! 🚢
